pkg/tracing/tracing.go

```diff
@@ -13,58 +18,269 @@ import (
 
 // Commandline options for exporter endpoints
 const (
-	jaegerEndpointOpt      = "jaeger-endpoint"
-	jaegerAgentEndpointOpt = "jaeger-agent-endpoint"
+	jaegerEndpointOpt       = "jaeger-endpoint"
+	jaegerAgentEndpointOpt  = "jaeger-agent-endpoint"
+	jaegerSamplerOpt        = "jaeger-sampler"
+	jaegerSampleFractionOpt = "jaeger-sample-fraction"
+)
+
+// JaegerSamplerType to indicate different Jaeger sampler type
+type JaegerSamplerType uint8
+
+// Sampler types in Jaeger
+const (
+	// 'Never' - Don't sample any trace
+	Never JaegerSamplerType = iota
+	// 'Always' - Sample every trace
+	Always
+	// 'Probabilistic' - Sample based on sample fraction
+	Probabilistic
 )
 
+// SamplerTypeToString returns string representation of sampler type.
+func SamplerTypeToString(samplerType JaegerSamplerType) string {
+	switch samplerType {
+	case Never:
+		return "never"
+	case Always:
+		return "always"
+	case Probabilistic:
+		return "probabilistic"
+	default:
+		return "unknown"
+	}
+}
+
+// DefaultSampleFraction - Default sample fraction. By default every 1 in 10
+// traces will be sampled.
+var DefaultSampleFraction = 0.1
+
+// JaegerTraceConfig holds Jaeger trace information
+type JaegerTraceConfig struct {
+	// Jaeger collector endpoint to which agent sends spans in batches
+	JaegerEndpoint string
+	// Jaeger agent endpoint to which spans are sent
+	JaegerAgentEndpoint string
+	// Jaeger sampler (0 - never or 1 - always or 2 - probabilistic)
+	JaegerSampler JaegerSamplerType
+	// Jaeger sample fraction to use in case sampler type is probabilistic
+	JaegerSampleFraction float64
+}
+
+// Jaeger exporter
+var jaegerExporter *jaeger.Exporter
+
+// Global Jaeger trace config
+var jaegerTraceConfig *JaegerTraceConfig
```

Comment:
as this is a global option. what happens if gd2 restarts?

don't we need to take lock as we are updating global options

Reply:
This global option is always modified under a transaction (under lock) as part of "enable", "update" and "delete" commands. When gd2 restarts on a node for example, the configuration is only read from etcd in the startup sequence and the in-memory config on that node is updated. Therefore, I don't see an issue with respect to this scenario as the node will get the latest configuration if trace config was update while the node was down. On the other hand, if config change happens just after it joins the etcd cluster, it will anyway get the updated configuration as part of the "enable", "update" or "delete" transaction.